lib/firmware/cosmos.c

```diff
@@ -0,0 +1,225 @@
+#include "trezor/crypto/segwit_addr.h"
+#include <stdbool.h>
+#include <time.h>
+
+static CONFIDENTIAL HDNode node;
+static SHA256_CTX ctx;
+static bool has_message;
+static bool initialized;
+static uint32_t address_n[8];
+static size_t address_n_count;
+static uint32_t msgs_remaining;
+static uint64_t sequence;
+
+bool cosmos_path_mismatched(const CoinType *_coin,
+                            const uint32_t *_address_n,
+                            const uint32_t _address_n_count)
+{
+    // m/44' : BIP44-like path
+    // m / purpose' / bip44_account_path' / account' / x / y
+    bool mismatch = false;
+    mismatch |= _address_n_count != 5;
+    mismatch |= _address_n_count > 0 && (_address_n[0] != (0x80000000 + 44));
+    mismatch |= _address_n_count > 1 && (_address_n[1] != _coin->bip44_account_path);
+    mismatch |= _address_n_count > 2 && (_address_n[2] & 0x80000000) == 0;
+    mismatch |= _address_n_count > 3 && _address_n[3] != 0;
+    mismatch |= _address_n_count > 4 && _address_n[4] != 0;
+    return mismatch;
+}
+
+/*
+ * Gets the address
+ *
+ * _node: HDNode from which the address is to be derived
+ * address: output buffer
+ *
+ * returns true if successful
+ */
+bool cosmos_getAddress(const HDNode *_node, char *address)
+{
+    uint8_t hash160Buf[RIPEMD160_DIGEST_LENGTH];
+    ecdsa_get_pubkeyhash(_node->public_key, HASHER_SHA2_RIPEMD, hash160Buf);
+
+    uint8_t fiveBitExpanded[RIPEMD160_DIGEST_LENGTH * 8 / 5];
+    size_t len = 0;
+    convert_bits(fiveBitExpanded, &len, 5, hash160Buf, 20, 8, 1);
+    return bech32_encode(address, "cosmos", fiveBitExpanded, len) == 1;
+}
+
+void sha256UpdateEscaped(SHA256_CTX *_ctx, const char *s, size_t len)
+{
+    while (len > 0)
+    {
+        if (s[0] == '"')
+        {
+            sha256_Update(_ctx, (uint8_t *)"\\\"", 2);
+        }
+        else if (s[0] == '\\')
+        {
+            sha256_Update(_ctx, (uint8_t *)"\\\\", 2);
+        }
+        else
+        {
+            sha256_Update(_ctx, (uint8_t *)&s[0], 1);
+        }
+        s = &s[1];
+        len--;
+    }
+}
+
+bool cosmos_signTxInit(const HDNode* _node,
+                       const uint32_t* _address_n,
+                       const size_t _address_n_count,
+                       const uint64_t account_number,
+                       const char *chain_id,
+                       const size_t chain_id_length,
+                       const uint32_t fee_uatom_amount,
+                       const uint32_t gas,
+                       const char *memo,
+                       const size_t memo_length,
+                       const uint64_t _sequence,
+                       const uint32_t msg_count)
+{
+
+    initialized = true;
+    msgs_remaining = msg_count;
+    has_message = false;
+    sequence = _sequence;
+
+    memzero(&node, sizeof(node));
+    memcpy(&node, _node, sizeof(node));
+    memzero(address_n, sizeof(address_n));
+    memcpy(address_n, _address_n, sizeof(address_n));
```

Comment:
Potential for UB here, if `address_n` and `_address_n` aren't the same length. I recommend:

checking that:

```
_address_n_count <= sizeof(address_n)/sizeof(address_n[0])
```

and:

```suggestion
    memcpy(address_n, _address_n, _address_n_count * sizeof(*address_n));
```